### PR TITLE
10935 Add Harmony Provider 

### LINKED ISSUE #10935: [ENHANCEMENT] Add Harmony AI Provider Support for GPT-OSS Models
### Problem (one or two sentences)

# Add Harmony AI Provider Support for GPT-OSS Models

### Overview
Add first-class support for **Harmony**, OpenAI's response format specification for open-weight GPT-OSS models (gpt-oss-20b and gpt-oss-120b).

### Problem
Currently, users wanting to use Harmony-compatible endpoints must use the generic "OpenAI Compatible" provider, which:
- Requires manual base URL configuration each time
- Lacks visibility as a dedicated provider option in the UI
- Doesn't clearly indicate Harmony-specific features like reasoning effort levels
- Provides no optimized configuration for Harmony's unique capabilities

### Solution
Implement Harmony as a **first-class API provider** in Roo Code, similar to how Groq, SambaNova, and other providers are implemented.

### Benefits
- **Better User Experience**: Harmony appears as a dedicated provider option in settings
- **Clear Configuration**: Pre-configured defaults for Harmony endpoints with customizable base URL
- **Feature Visibility**: Reasoning effort levels (low/medium/high) become discoverable and configurable
- **Reliability**: Extends proven BaseOpenAiCompatibleProvider pattern for consistency
- **Open-Weight Support**: Enables developers to use open-source models with Roo Code

### What Will Be Implemented

**Models:**
- `gpt-oss-20b` - 20B parameter model, optimized for speed (default)
- `gpt-oss-120b` - 120B parameter model, optimized for quality

**Features:**
- 128,000 token context window for comprehensive codebase analysis
- Reasoning effort levels: low, medium, high
- Streaming support for real-time responses
- Function/tool calling capabilities
- Configurable API key and base URL
- Full i18n support (17 languages)

**Code Quality:**
- Complete test coverage (12 test cases)
- Comprehensive JSDoc documentation
- TypeScript strict mode compliance
- Follows Roo Code contribution guidelines

### References
- [OpenAI Harmony Specification](https://developers.openai.com/cookbook/articles/openai-harmony)
- [Harmony GitHub Repository](https://github.com/openai/harmony)
- [Roo Code Contributing Guidelines](https://github.com/RooCodeInc/Roo-Code/blob/main/CONTRIBUTING.md)

### Alignment with Project Goals
- ✅ **Reliability First**: Extends proven provider pattern for consistency
- ✅ **Enhanced UX**: First-class provider option with clear feature visibility
- ✅ **Agent Performance**: Support for reasoning effort levels enables better model configuration

### Context (who is affected and when)

Roo Code users when using a GPT-OSS Harmony compatible endpoint with OpenAI Compatible provider

### Desired behavior (conceptual, not technical)

- [ ] Harmony provider is available as a selectable option in API provider settings
- [ ] Both gpt-oss-20b and gpt-oss-120b models are selectable
- [ ] Configurable API key and base URL with sensible defaults
- [ ] Reasoning effort selector (low/medium/high) is visible in settings

### Constraints / preferences (optional)

- ✅ **Reliability First**: Extends proven provider pattern for consistency
- ✅ **Enhanced UX**: First-class provider option with clear feature visibility
- ✅ **Agent Performance**: Support for reasoning effort levels enables better model configuration

- Full i18n support (17 languages)

**Code Quality:**
- Complete test coverage (12 test cases)
- Comprehensive JSDoc documentation
- TypeScript strict mode compliance
- Follows Roo Code contribution guidelines

### Request checklist

- [x] I've searched existing Issues and Discussions for duplicates
- [x] This describes a specific problem with clear context and impact

### Roo Code Task Links (optional)

_No response_

### Acceptance Criteria
- [ ] Harmony provider is available as a selectable option in API provider settings
- [ ] Both gpt-oss-20b and gpt-oss-120b models are selectable
- [ ] Configurable API key and base URL with sensible defaults
- [ ] Reasoning effort selector (low/medium/high) is visible in settings
- [ ] All tests pass (npm test)
- [ ] TypeScript compilation succeeds with no errors
- [ ] README documents Harmony as a supported provider
- [ ] JSDoc documentation is comprehensive

### Proposed approach (optional)

Add as Provider: 
        modified:   README.md
        modified:   packages/types/src/provider-settings.ts
        new file:   packages/types/src/providers/harmony.ts
        modified:   packages/types/src/providers/index.ts
        modified:   src/api/index.ts
        new file:   src/api/providers/__tests__/harmony.spec.ts
        new file:   src/api/providers/harmony.ts
        modified:   src/api/providers/index.ts
        new file:   test-harmony-api.ts
        modified:   webview-ui/src/components/settings/ApiOptions.tsx
        new file:   webview-ui/src/components/settings/providers/Harmony.tsx
        modified:   webview-ui/src/components/settings/providers/index.ts
        modified:   webview-ui/src/components/ui/hooks/useSelectedModel.ts
        modified:   webview-ui/src/i18n/locales/en/settings.json

### Trade-offs / risks (optional)

None

### PR DESCRIPTION
**Pull request by ajgreyling**
### Related GitHub Issue

Closes: #10935

### Description

This PR fixes vLLM compatibility issues with the Harmony provider by addressing two key problems:

1. **Removing unsupported `strict` parameter**: vLLM doesn't support OpenAI's `strict` mode parameter for function calling. The base class adds this parameter, but Harmony (which uses vLLM) needs it removed to prevent warnings and potential errors.

2. **Preserving optional parameters**: The base class's `convertToolSchemaForOpenAI` method converts all properties to required fields for OpenAI strict mode. However, vLLM requires that optional parameters (especially those with nullable types like `["string", "null"]`) remain optional. This is critical for tools like `browser_action` that have optional parameters.

**Implementation details:**

- **`convertToolsForOpenAI` override**: Removes the `strict` parameter from all function tools after conversion, ensuring vLLM compatibility.
- **`convertToolSchemaForOpenAI` override**: 
  - Preserves the original `required` array structure
  - Converts nullable types (e.g., `["string", "null"]`) to non-nullable types (e.g., `"string"`)
  - Excludes nullable properties from the `required` array, keeping them optional
  - Recursively processes nested objects and arrays
  - Ensures `additionalProperties: false` is set (required by OpenAI Responses API)

This fix ensures that tools with optional parameters work correctly with vLLM backends while maintaining compatibility with the OpenAI-compatible API format.

### Test Procedure

**Unit Tests Added:**

1. **`convertToolsForOpenAI` tests:**
   - Verifies `strict` parameter is removed from function tools
   - Ensures all other tool properties are preserved
   - Tests MCP tool handling (which have `strict: false` from parent)
   - Verifies non-function tools pass through unchanged
   - Tests undefined input handling

2. **Optional parameter preservation test:**
   - Tests `browser_action` tool with nullable optional parameters (`url`, `coordinate`)
   - Verifies nullable types are converted to non-nullable
   - Ensures only required fields (`action`) are in the `required` array
   - Confirms optional parameters remain optional for vLLM compatibility

**Running tests:**

```bash
npx vitest run src/api/providers/__tests__/harmony.spec.ts
```

All existing tests continue to pass, and the new tests verify the vLLM compatibility fixes.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to GitHub Issue #10935.
- [x] **Scope**: Changes are focused on fixing vLLM compatibility issues with the Harmony provider.
- [x] **Self-Review**: Code has been reviewed for correctness and follows existing patterns.
- [x] **Testing**: Comprehensive unit tests have been added to cover the new functionality.
- [x] **Documentation Impact**: No user-facing documentation changes required (internal implementation fix).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

N/A - Backend implementation fix with no UI changes.

### Documentation Updates

- [x] No documentation updates are required.

This is an internal implementation fix that improves compatibility with vLLM backends. No user-facing documentation changes are needed.

### Additional Notes

**Key design decisions:**

1. **Why override instead of modifying base class**: The base class behavior is correct for OpenAI's strict mode, but Harmony uses vLLM which doesn't support strict mode. Overriding allows Harmony-specific behavior without affecting other providers.

2. **Nullable type handling**: The implementation converts `["string", "null"]` to `"string"` while tracking which properties were nullable. This ensures the `required` array only includes truly required fields, not optional nullable ones.

3. **Backward compatibility**: The changes maintain full backward compatibility with existing tool schemas while fixing the vLLM-specific issues.

**Testing considerations:**

- Tests cover both regular tools and MCP tools
- Tests verify the critical `browser_action` use case with nullable optional parameters
- All edge cases (undefined, non-function tools, nested objects) are covered

### Get in Touch

Contact me on Discord @ ajgreyling

----

> [!IMPORTANT]
> Adds Harmony provider to handle GPT-OSS models with vLLM compatibility, including new handler, settings, and tests.
> 
>   - **Behavior**:
>     - Adds `HarmonyHandler` in `harmony.ts` to handle GPT-OSS models with vLLM compatibility.
>     - Removes `strict` parameter in `convertToolsForOpenAI()` to prevent vLLM warnings.
>     - Preserves optional parameters in `convertToolSchemaForOpenAI()`.
>   - **Files**:
>     - Adds `harmony.ts` to `src/api/providers/`.
>     - Updates `index.ts` in `src/api/` and `src/api/providers/` to include `HarmonyHandler`.
>     - Modifies `provider-settings.ts` to add Harmony provider settings.
>   - **Tests**:
>     - Adds `harmony.spec.ts`, `harmony-edge-cases.spec.ts`, and `harmony-roo-code-integration.spec.ts` for unit and integration tests.
>     - Includes tests for `convertToolsForOpenAI()` and optional parameter handling.
>   - **UI**:
>     - Updates `ApiOptions.tsx` and `constants.ts` to include Harmony provider in settings.
>     - Adds `Harmony.tsx` for Harmony provider settings UI.
>   - **Misc**:
>     - Updates `README.md` to fix formatting.
>     - Updates `package.json` to a new `pnpm` version.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 698cd2332966e49999f26d27d38e65a7c1b53e5c. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by ghost** [comment]
> **Rooviewer**  Clock &nbsp;&nbsp;[See task on Roo Cloud](https://app.roocode.com/cloud-jobs/B0wPlA3W?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review)

Review complete. Found 3 issues that should be addressed before merging:

- [ ] Protected method `convertToolsForOpenAI` called directly in tests - use a test wrapper class pattern (see Cerebras tests for example)
- [ ] Development/debug files (`test-harmony-api.ts`, `test_harmony_toolcall_fix.py`) placed in project root - consider removing or relocating
- [ ] Unrelated `packageManager` version change in `package.json` - consider reverting or submitting separately

<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=10936&commentId=3794513585).</sub>

File: src/api/providers/__tests__/harmony.spec.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+					function: {
+						name: "test_tool",
+						description: "A test tool",
+						parameters: { type: "object", properties: {} },
+						strict: true, // This will be added by parent class
+					},
+				},
+			]
+
+			const converted = handler.convertToolsForOpenAI(tools)
```
Comment (ghost) [inline src/api/providers/__tests__/harmony.spec.ts:138]: The `convertToolsForOpenAI` method is `protected` in the parent class, so calling it directly on the handler instance will cause TypeScript compilation errors. The Cerebras tests demonstrate the correct pattern: create a test wrapper class that exposes the protected method via a public method. For example:

```typescript
class TestHarmonyHandler extends HarmonyHandler {
    public testConvertToolsForOpenAI(tools: any[] | undefined) {
        return this.convertToolsForOpenAI(tools)
    }
}
```

Then use `testHandler.testConvertToolsForOpenAI(tools)` in tests.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10936&fixId=fd6d910e8f) or mention @roomote and request a fix.</sub>

File: test-harmony-api.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import OpenAI from "openai"
```
Comment (ghost) [inline test-harmony-api.ts:1]: This development/debug script and the accompanying `test_harmony_toolcall_fix.py` are placed in the project root. Consider either removing these files from the PR (they appear to be local development tools) or moving them to an appropriate location such as `scripts/` or a documentation directory if they serve as examples.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10936&fixId=79c1bf3db1) or mention @roomote and request a fix.</sub>

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "roo-code",
-	"packageManager": "pnpm@10.8.1",
+	"packageManager": "pnpm@10.28.1",
 	"engines": {
```
Comment (ghost) [inline package.json:4]: This `packageManager` version change from `pnpm@10.8.1` to `pnpm@10.28.1` appears unrelated to the Harmony provider feature. Consider reverting this change or submitting it as a separate PR to keep the feature scope focused.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10936&fixId=b33f6f1d86) or mention @roomote and request a fix.</sub>
